torch/csrc/api/include/torch/data/ctf/ctf_parser.hpp

```diff
@@ -0,0 +1,223 @@
+#pragma once
+
+#include "torch/data/ctf/reader.hpp"
+#include "torch/csrc/utils/disallow_copy.h"
+
+#include <map>
+#include <vector>
+#include <cassert>
+#include <iostream>
+#include <ostream>
+#include <stdint.h>
+
+/*
+ * CTF general format
+ * [Sequence_Id](Sample or Comment)+
+ *   where
+ *          sequence_Id=(empty|[0-9]+)
+ *          Sample=|Input_Name (Value )*
+ *          Comment=|# some content
+ *
+ * 100 |a 1 2 3 |b 100 200
+ * 100 |a 4 5 6 |b 101 201
+ * 100 |b 102983 14532 |a 7 8 9
+ * 100 |a 7 8 9
+ * 200 |b 300 400 |a 10 20 30
+ * 333 |b 500 100
+ * 333 |b 600 -900
+ * 400 |a 1 2 3 |b 100 200
+ * |a 4 5 6 |b 101 201
+ * |a 4 5 6 |b 101 201
+ * 500 |a 1 2 3 |b 100 200
+ */
+
+/* General use string aprsing helpers */
+static const char SPACE_CHAR = ' ';
+static const char TAB_CHAR = '\t';
+static const char NAME_PREFIX = '|';
+static const char INDEX_DELIMITER = ':';
+static const char ROW_DELIMITER = '\n';
+static const char ESCAPE_SYMBOL = '#';
+
+inline bool isNamePrefix(const char &c)
+{
+    return (c == NAME_PREFIX);
+}
+
+inline bool isCommentPrefix(const char &c)
+{
+    return (isNamePrefix(c));
+}
+
+inline bool isCommentSuffix(const char &c)
+{
+    return (c == '#');
+}
+
+inline bool isDecimalPoint(const char &c)
+{
+    return (c == '.');
+}
+
+inline bool isSparseValueDelimiter(const char &c)
+{
+    return (c == ':');
+}
+
+inline bool isDigit(const char &c)
+{
+    return (c >= '0' && c <= '9');
+}
+
+inline bool isAlpha(const char &c)
+{
+    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
+}
+
+inline bool isSign(const char &c)
+{
+    return c == '+' || c == '-';
+}
+
+inline bool isNumber(const char &c)
+{
+    return (isDigit(c) || isDecimalPoint(c) || isSign(c));
+}
+
+inline bool isPrintable(const char &c)
+{
+    return c >= SPACE_CHAR;
+}
+
+inline bool isNonPrintable(const char &c)
+{
+    return !isPrintable(c);
+}
+
+inline bool isValueDelimiter(const char &c)
+{
+    return c == SPACE_CHAR || c == TAB_CHAR;
+}
+
+inline bool isColumnDelimiter(const char &c)
+{
+    return isValueDelimiter(c) || (isNonPrintable(c) && c != ROW_DELIMITER);
+}
+
+inline bool isEOL(const char &c)
+{
+    return (c == '\r' || c == '\n');
+}
+
+inline bool isEscapeDelimiter(const char &c)
+{
+    return (c == '\'' || c == '"');
+}
+
+/* CTF Types */
+enum CTFValueType
+{
+    Unknown = 0x0,
+    Float = 0x1,
+    Double = 0x2,
+    Float16 = 0x3,
+    Int8 = 0x4,
+    Int16 = 0x5
+};
+static const std::string ctf_value_type_str[] = {"Unknown", "Float", "Double", "Float16", "Int8", "Int16"};
+
+typedef size_t CTFSequenceID;
+typedef std::string CTFName;
+typedef std::string CTFComment;
+
+struct CTFValue
+{
+    explicit CTFValue() : type(CTFValueType::Unknown),
+                          value(0),
+                          index(SIZE_MAX){}; // TODO: Kill copy assignment/copy constructor?
```

Comment:
That was just a memo for the future me to think about performance gains and best practices. Until now, I don think there is a good reason to prevent copy constructors for structs

Reply:
I will remove the comments in the final version